draft-ietf-httpbis-cache-digest.md

```diff
@@ -161,85 +159,147 @@ When generating CACHE_DIGEST, a client MUST NOT include cached responses whose U
+4. Let `key` be the return value of {{key}} with `URL` and `ETag` as inputs.
+5. Let `h1` be the return value of {{hash}} with `key` and `N` as inputs.
+6. Let `fingerprint` be the return value of {{fingerprint}} with `key`, `N` and `f` as inputs.
+7. Let `h2` be the return value of {{hash}} with `fingerprint` and `N` as inputs, XORed with `h1`.
+8. Let `h` be `h1`.
+9. Let `position_start` be 40 + `h` * `f`.
+10. Let `position_end` be `position_start` + `f` * `b`.
+11. While `position_start` < `position_end`:
+    1. Let `bits` be `f` bits from `digest_value` starting at `position_start`.
+    2. If `bits` is `fingerprint`, set `bits` to all zeros and terminate these steps.
+    3. Add `f` to `position_start`.
+12. If `h` is not `h2`, set `h` to `h2` and return to step 9.
+
+### Computing a fingerprint value {#fingerprint}
 
-Given:
+Given the following inputs:
 
-* `URL`, an array of characters
-* `ETag`, an array of characters
-* `validators`, a boolean
+* `key`, an array of characters
 * `N`, an integer
-* `P`, an integer
+* `f`, an integer indicating the number of output bits
 
-`hash-value` can be computed using the following algorithm:
+1. Let `hash-value` be the return value of {{hash}} with `key` and `N` as inputs.
+2. Let `h` be the number of bits in `hash-value`
+3. Let `fingerprint-value` be 0
+4. While `fingerprint-value` is 0 and `h` > `f`:
+    4.1. Let `fingerprint-value` be the `f` least significant bits of `hash-value`.
+    4.2. Let `hash-value` be the the `h`-`f` most significant bits of `hash-value`.
+    4.3. `h` -= `f`
+5. If `fingerprint-value` is 0, let `fingerprint-value` be 1.
+6. Return `fingerprint-value`.
+
+Note: Step 5 is to handle the extremely unlikely case where a SHA-256 digest of `key` is all zeros. The implications of it means that
+there's an infitisimaly larger probability of getting a `fingerprint-value` of 1 compared to all other values. This is not a problem for any
+practical purpose.
+
+
+
+### Computing the key {#key}
+
+Given the following inputs:
+
+* `URL`, an array of characters
+* `ETag`, an array of characters
 
 1. Let `key` be `URL` converted to an ASCII string by percent-encoding as appropriate {{RFC3986}}.
-2. If `validators` is true and `ETag` is not null:
+2. If `ETag` is not null:
    1. Append `ETag` to `key` as an ASCII string, including both the `weak` indicator (if present) and double quotes, as per {{RFC7232}}, Section 2.3.
-3. Let `hash-value` be the SHA-256 message digest {{RFC6234}} of `key`, expressed as an integer.
-4. Truncate `hash-value` to log2( `N` \* `P` ) bits.
+3. Return `key`
+
+### Computing a Hash Value {#hash}
 
+Given the following inputs:
+
+* `key`, an array of characters.
+* `N`, an integer
+
+`hash-value` can be computed using the following algorithm:
+
+1. Let `hash-value` be the SHA-256 message digest {{RFC6234}} of `key`, expressed as an integer.
+2. Return `hash-value` modulo N.
```

Comment:
This is difficult to do in JavaScript where uint operations are typically still limited to 32 bits. The truncation in the previous proposal (step 4) is more compatible and, if I understand correctly, achieves the same objective. Can this be changed to something that does not require 512 bit integer modulo?

Reply:
I wonder if the need for an integer modulo is due to an error in the specification.

The text in the PR states that `N` is `a prime number smaller than 2\*\*32`. Could it be the case that N is something to be defined as 2<sup>N</sup>?

If that is the case, the modulo operation can be implemented by using bitwise AND.

Reply:
OK, I'll truncate the hash before the modulo operation.